README.md
# ghci-daemon
This package allows you to wrap a "stack ghci" command and control it by sending
commands to a network socket (or using the `ghci-client` executable).

## How to use?
Start the daemon:
```
ghci-daemon
```

Send a command:
```
ghci-client -c ":reload"
```

(If you installed this package using stack you might want to use `stack exec`
to call the executables.)



ChangeLog.md
# Changelog for ghci-daemon

## Unreleased changes


src/Language/Haskell/GhciDaemon/Client.hs
module Language.Haskell.GhciDaemon.Client
    ( Settings(..)
    , Client(..)
    , mkClient
    , runClient
    , closeClient
    , exec
    , ghciClient
    ) where

import           Control.Monad (when)
import           Control.Monad.Loops (untilM_)
import           Control.Exception (bracket)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BC
import           Data.Semigroup ((<>))
import           Options.Applicative
import           Network.Socket hiding (send, sendTo, recv, recvFrom)
import           Network.Socket.ByteString (sendAll, recv)
import           System.IO (BufferMode(..), stdin, hIsEOF, hSetBuffering)

-- |CLI Options
data Settings = Settings
    { port :: Int
    , repl :: Bool
    , cmd  :: String
    , host :: String
    }

-- |Everything needed to send a request to the server
data Client = Client
    { cSettings :: Settings
    , cSocket   :: Socket
    }

settings :: Parser Settings
settings = Settings
    <$> option auto
         ( long "port"
        <> short 'p'
        <> help "The port to connect to"
        <> metavar "PORT"
        <> value 3497)
    <*> switch
         ( long "repl"
        <> short 'r'
        <> help "Start a REPL")
    <*> strOption
         ( long "cmd"
        <> short 'c'
        <> help "Execute command"
        <> metavar "CMD"
        <> value "")
    <*> strOption
         ( long "host"
        <> short 'h'
        <> help "Host to connect to"
        <> metavar "HOSTNAME"
        <> value "127.0.0.1")

-- |Open the socket
openSocket :: Settings -> IO Socket
openSocket s = do
    addr:_ <- getAddrInfo (Just hints) (Just . host $ s) (Just . show . port $ s)
    sock   <- socket (addrFamily addr) (addrSocketType addr) (addrProtocol addr)
    connect sock $ addrAddress addr
    return sock
        where
            hints = defaultHints { addrSocketType = Stream }

-- |Parse all cli args and open socket
mkClient :: IO Client
mkClient = do
    s <- execParser opts
    sock <- openSocket s
    return $ Client s sock
        where
            opts = info (settings <**> helper)
                 ( fullDesc
                <> progDesc "Connect to a ghci daemon"
                <> header "ghci-client") 

-- |Execute a Client
runClient :: Client -> IO ()
runClient c = do
    when ((cmd . cSettings $ c) /= "") $
        exec (cSocket c) (cmd . cSettings $ c)
    when (repl . cSettings $ c) $ do
        hSetBuffering stdin NoBuffering
        putStrLn "Starting remote repl"
        runRepl (cSocket c)

-- |Send a command to server and print result to stdout
exec :: Socket -> String -> IO ()
exec s c = do
    sendAll s (BC.pack c)
    recv s 4096 >>= BC.putStrLn

-- |Run a basic repl
runRepl :: Socket -> IO ()
runRepl s = prompt `untilM_` (hIsEOF stdin)
    where
        prompt = do
            c <- getLine
            exec s c

-- |Parse and execute all cli arguments
ghciClient :: IO ()
ghciClient = do
    bracket mkClient closeClient runClient

-- |Close connection to server
closeClient :: Client -> IO ()
closeClient Client {cSocket = sock, cSettings = _} = close sock


src/Language/Haskell/GhciDaemon/Server.hs
{-# LANGUAGE OverloadedStrings #-}
module Language.Haskell.GhciDaemon.Server
    ( Daemon(..)
    , Settings(..)
    , mkDaemon
    , runDaemon
    , killDaemon
    , ghciDaemon
    ) where

import           Control.Exception (bracket)
import           Control.Monad (forever)
import qualified Data.ByteString.Char8 as BC
import           Data.Semigroup ((<>))
import           Language.Haskell.Ghcid (Ghci, startGhci, exec, interrupt)
import           Network.Socket hiding (send, sendTo, recv, recvFrom)
import           Network.Socket.ByteString (send, recv)
import           Options.Applicative
import           System.Posix.Daemon (Redirection(DevNull), runDetached)

-- |Options passed to the Server (e.g as cli arguments)
data Settings = Settings
    { port     :: Int
    , cmd      :: String
    , nofork   :: Bool
    }

-- |All information needed to run the server
data Daemon = Daemon
    { dSettings :: Settings
    , dSocket   :: Socket
    , dGhci     :: Ghci
    }

settings :: Parser Settings
settings = Settings 
    <$> option auto
         ( long "port"
        <> short 'p'
        <> help "The port for the server to listen to"
        <> metavar "PORT"
        <> value 3497)
    <*> strOption
         ( long "cmd"
        <> short 'c'
        <> help "The command to run ghci"
        <> metavar "GHCICMD"
        <> value "stack ghci")
    <*> switch
         ( long "no-fork"
        <> short 'n'
        <> help "Do not run in background")

ghciDaemon :: IO ()
ghciDaemon = do
    d <- mkDaemon
    let act = bracket (return d) killDaemon runDaemon
    if nofork . dSettings $ d
        then act
        else runDetached Nothing DevNull act

-- |Create a daemon
mkDaemon :: IO Daemon
mkDaemon = do
    s <- execParser opts
    sock <- openSocket s
    (ghci, msg) <- startGhci (cmd s) Nothing (\_ -> putStrLn)
    print msg
    return (Daemon s sock ghci)
        where
            opts = info (settings <**> helper)
                 ( fullDesc
                <> progDesc "Run a ghci daemon"
                <> header "ghci-daemon")

-- |Run a daemon
runDaemon :: Daemon -> IO ()
runDaemon (Daemon _ sock ghci) = forever $ do
    (conn, _) <- accept sock
    msg <- BC.unpack <$> recv conn 4096
    res <- exec ghci msg
    _ <- send conn (BC.pack . unlines $ res)
    close conn

-- |Try to stop gracefully. Kill ghci if this does not succeed.
killDaemon :: Daemon -> IO ()
killDaemon (Daemon _ sock ghci) = do
    close sock
    interrupt ghci

-- |Open the network socket
openSocket :: Settings -> IO Socket
openSocket s = do
    addr:_ <- getAddrInfo (Just hints) Nothing (Just . show $ port s)
    sock <- socket (addrFamily addr) (addrSocketType addr) (addrProtocol addr)
    setSocketOption sock ReuseAddr 1
    bind sock (addrAddress addr)
    listen sock 1
    return sock
        where
            hints = defaultHints {
                 addrFlags = [AI_PASSIVE]
               , addrSocketType = Stream
               }
